api/agent/drivers/docker/docker.go

```diff
@@ -102,18 +105,88 @@ func NewDocker(conf drivers.Config) *DockerDriver {
 		}
 	}
 
-	if conf.DockerLoadFile != "" {
-		err = loadDockerImages(driver, conf.DockerLoadFile)
-		if err != nil {
-			logrus.WithError(err).Fatalf("cannot load docker images in %s", conf.DockerLoadFile)
-		}
+	err = checkDockerVersion(ctx, driver)
+	if err != nil {
+		logrus.WithError(err).Fatal("docker version error")
+	}
+
+	// start the cleanup as early as possible
+	go containerCleaner(ctx, driver)
+
+	// before we do anything else, let's pre-load requested images
+	err = loadDockerImages(ctx, driver)
+	if err != nil {
+		logrus.WithError(err).Fatalf("cannot load docker images in %s", conf.DockerLoadFile)
+	}
+
+	// finally spawn pool if enabled
+	if conf.PreForkPoolSize != 0 {
+		driver.pool = NewDockerPool(conf, driver)
 	}
 
 	return driver
 }
 
-func checkDockerVersion(driver *DockerDriver, expected string) error {
-	info, err := driver.docker.Info(context.Background())
+// containerCleaner scans and destroys previously left over containers that were managed
+// by this docker driver. This operation is executed once and if it fails, it will not
+// retry the procedure.
+func containerCleaner(ctx context.Context, driver *DockerDriver) {
+
+	// Label Tag is used to isolate this cleanup. If docker has other containers
+	// that are not managed by fn-agent, then this tag can make sure those containers
+	// are not killed. For this reason, we require this tag to be set.
+	if driver.conf.ContainerLabelTag == "" {
+		return
+	}
+
+	filter := fmt.Sprintf("%s=%s", FnAgentClassifierLabel, driver.conf.ContainerLabelTag)
+
+	opts := docker.ListContainersOptions{
+		All: true, // let's include containers that are not running, but not destroyed
+		Filters: map[string][]string{
+			"label": []string{filter},
+		},
+		Context: ctx,
+	}
+
+	containers, err := driver.docker.ListContainers(opts)
+	if err != nil {
+		logrus.WithError(err).Errorf("cannot list docker containers with filter %s", filter)
+		return
+	}
+
+	for _, item := range containers {
+		logrus.Errorf("checking %+v", item)
```

Comment:
nit: can we log this with debug level? it's just alarming I guess for someone not familiar to see a bunch of errors fly out at the beginning that aren't really errors and we're logging terminations below already